karabo_data/geometry2.py

```diff
@@ -500,6 +493,128 @@ def to_distortion_array(self):
 
         return distortion
 
+    def _tile_slice(self, tileno):
+        """Implement in subclass: which part of module array each tile is.
+        """
+        raise NotImplementedError
+
+    def _module_coords_to_tile(self, slow_scan, fast_scan):
+        """Implement in subclass: positions in module to tile numbers & pos in tile
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def _adjust_pixel_coords(cls, ss_coords, fs_coords):
+        """Called by get_pixel_positions; overridden by DSSC"""
+        pass
+
+    def get_pixel_positions(self, centre=True):
+        """Get the physical coordinates of each pixel in the detector
+
+        The output is an array with shape like the data, with an extra dimension
+        of length 3 to hold (x, y, z) coordinates. Coordinates are in metres.
+
+        If centre=True, the coordinates are calculated for the centre of each
+        pixel. If not, the coordinates are for the first corner of the pixel
+        (the one nearest the [0, 0] corner of the tile in data space).
+        """
+        out = np.zeros(self.expected_data_shape + (3,), dtype=np.float64)
+
+        # Prepare some arrays to use inside the loop
+        pixel_ss_coord, pixel_fs_coord = np.meshgrid(
+            np.arange(0, self.frag_ss_pixels, dtype=np.float64),
+            np.arange(0, self.frag_fs_pixels, dtype=np.float64),
+            indexing='ij'
+        )
+
+        # This is where the DSSC subclass shifts odd rows by half a pixel
+        self._adjust_pixel_coords(pixel_ss_coord, pixel_fs_coord)
+
+        if centre:
+            # A pixel is from n to n+1 in each axis, so centres are at n+0.5.
+            pixel_ss_coord += 0.5
+            pixel_fs_coord += 0.5
+
+        for m, mod in enumerate(self.modules, start=0):
+            for t, tile in enumerate(mod, start=0):
+                corner_x, corner_y, corner_z = tile.corner_pos * self.pixel_size
+                ss_unit_x, ss_unit_y, ss_unit_z = tile.ss_vec * self.pixel_size
+                fs_unit_x, fs_unit_y, fs_unit_z = tile.fs_vec * self.pixel_size
```

Comment:
I do consider a box around the pixel, so if you pass `centre=False`, it will give a point just outside the pixel (the corner of the bounding box). I'm OK with that. I'm planning to implement the distortion array method based on this, and that uses offsets from the corner of the bounding box.

However, I have made a mistake and got the centres slightly off. We should be adding 2/3 instead of 1/2 in the slow-scan direction - for the same reason as the 4/3 multiplication in the other PR you reviewed recently.

Reply:
To illustrate what I meant about the centres being off, here's the plot before the correction:

![pixel-centres-before2](https://user-images.githubusercontent.com/327925/59913836-e4c4c880-9410-11e9-8243-17377490c3db.png)

And here it is after replacing the 1/2 shift with 2/3:

![pixel-centres-after](https://user-images.githubusercontent.com/327925/59913853-eee6c700-9410-11e9-8e0d-1f4d769c5e77.png)